### PR TITLE
consul/connect: fix regression where client connect images ignored

### LINKED ISSUE #9618: [bug/regression] nomad 1.0.0 ignores nomad meta.connect.(sidecar|gateway)_image
**Nomad version: 1.0.0**

The envoy proxy config set in the client meta are ignored since the introduction of the dynamic envoy version detection. This is not just an issue to configure a more recent envoy version but also because it now by default pulls envoy containers from the docker hub with rate limits.

This is the output from a `nomad nodes status -self -verbose` and has been unchanged since the 1.0.0 upgrade.
```
Meta
connect.gateway_image     = envoyproxy/envoy:v1.16.1
connect.log_level         = info
connect.proxy_concurrency = 1
connect.sidecar_image     = envoyproxy/envoy:v1.16.1
```
And after the nomad  upgrade planning the `countdash` nomad job:
```
  +/- Task: "connect-proxy-count-dashboard" (forces create/destroy update)
    +/- Config {
          args[0]: "-c"
          args[1]: "${NOMAD_SECRETS_DIR}/envoy_bootstrap.json"
          args[2]: "-l"
          args[3]: "${meta.connect.log_level}"
      +/- args[4]: "--disable-hot-restart" => "--concurrency"
      +   args[5]: "${meta.connect.proxy_concurrency}"
      +   args[6]: "--disable-hot-restart"
      +/- image:   "${meta.connect.sidecar_image}" => "envoyproxy/envoy:v${NOMAD_envoy_version}"
        }
```
And the associated docker container:
```
9de76a58cd4b        envoyproxy/envoy:v1.16.0                      "/docker-entrypoint.…"   13 minutes ago      Up 13 minutes                           connect-proxy-count-dashboard-82594608-0a7e-4e8c-8000-2aa5f6ccdd59
```

### PR DESCRIPTION
**Pull request by shoenig**
Nomad v1.0.0 introduced a regression where the client configurations
for `connect.sidecar_image` and `connect.gateway_image` would be
ignored despite being set. This PR restores that functionality.

There was a missing layer of interpolation that needs to occur for
these parameters. Since Nomad 1.0 now supports dynamic envoy versioning
through the ${NOMAD_envoy_version} psuedo variable, we basically need
to first interpolate
```
  ${connect.sidecar_image} => envoyproxy/envoy:v${NOMAD_envoy_version}
```
then use Consul at runtime to resolve to a real image, e.g.
```
  envoyproxy/envoy:v${NOMAD_envoy_version} => envoyproxy/envoy:v1.16.0
```
Of course, if the version of Consul is too old to provide an envoy
version preference, we then need to know to fallback to the old
version of envoy that we used before.
```
  envoyproxy/envoy:v${NOMAD_envoy_version} => envoyproxy/envoy:v1.11.2@sha256:a7769160c9c1a55bb8d07a3b71ce5d64f72b1f665f10d81aa1581bc3cf850d09
```
Beyond that, we also need to continue to support jobs that set the
sidecar task themselves, e.g.
```
  sidecar_task { config { image: "custom/envoy" } }
```
which itself could include the pseudo envoy version variable.

Fixes #9618

### CODE REVIEW

**Comment by shoenig** [comment]
Spot check every combination. 

From lowest to highest precedence: 
`consul version`: consul_old (envoy v1.11.2), consul_new (envoy v1.16.0)
`client.meta`: meta_unset, meta_set (envoy v1.14.4)
`sidecar_task.image`: image_unset, image_set (envoy v1.15.3)

```
Consul      meta.       image.         Expect    OK
----------------------------------------------------
consul_old  meta_unset  image_unset -> v1.11.2   [ ]
consul_old  meta_unset  image_set   -> v1.15.3   [ ]
consul_old  meta_set    image_unset -> v1.14.4   [ ]
consul_old  meta_set    image_set   -> v1.15.3   [ ]
consul_new  meta_unset  image_unset -> v1.16.0   [ ]
consul_new  meta_unset  image_set   -> v1.15.3   [ ]
consul_new  meta_set    image_unset -> v1.14.4   [ ]
consul_new  meta_set    image_set   -> v1.15.3   [ ]

```

File: client/allocrunner/taskrunner/envoy_version_hook_test.go [hunk 1]
```diff
@@ -96,6 +105,54 @@ func TestEnvoyVersionHook_tweakImage(t *testing.T) {
 	})
 }
 
+func TestEnvoyVersionHook_interpolateImage(t *testing.T) {
```
Comment (tgross) [inline client/allocrunner/taskrunner/envoy_version_hook_test.go:108]: Solid test coverage here 👍

**Review by tgross — APPROVED** [review]
LGTM

**Comment by HarryHarcourt** [comment]
We are running Nomad v1.0.2
And either this does not work, or we are trying to set the value in the wrong place.
Any help would be appreciated, we are putting this in our client agent configuration:
options {
  meta.connect.sidecar_image = "artifactory.standard.com/docker/envoy:1.11.2"
}
But do not see any change when running nomad node status -self -verbose
Meta
connect.gateway_image     = envoyproxy/envoy:v${NOMAD_envoy_version}
connect.log_level         = info
connect.proxy_concurrency = 1
connect.sidecar_image     = envoyproxy/envoy:v${NOMAD_envoy_version}

**Comment by tgross** [comment]
Hi @HarryHarcourt! I think you want:

```hcl
client {
  meta {
    "connect.sidecar_image" = "artifactory.standard.com/docker/envoy:1.11.2"
  }
}
```

But if that doesn't work for you please open a new issue. By the way, the [`options`](https://www.nomadproject.io/docs/configuration/client#options-parameters) block is deprecated since Nomad 0.9 and IIRC it doesn't get the new configuration fields.

**Comment by HarryHarcourt** [comment]
Confirmed that this works.
If this is moving to an options approach, what plugin would this be configured against?  There is no documentation to support this approach.

**Comment by idrennanvmware** [comment]
@HarryHarcourt its just the standard nomad "client" configuration.   Not specific plugins or drivers around them.

```
{
    "client": {
        "enabled": true,
        "network_interface" : "???",
        "meta": "connect.sidecar_image": "{{ default_docker_registry }}/{{ images_dict_data.envoy }}", "connect.gateway_image": "{{ default_docker_registry }}/{{ images_dict_data.envoy }}"
        }]
    }
}
```

Please ignore our templating lol

**Pull request merged**